crates/ruff_python_parser/src/invalid.rs

```diff
@@ -0,0 +1,810 @@
+/*!
+Defines some helper routines for rejecting invalid Python programs.
+
+These routines are named in a way that supports qualified use. For example,
+`invalid::assignment_targets`.
+*/
+
+use lalrpop_util::ParseError as LalrpopError;
+
+use {ruff_python_ast::Expr, ruff_text_size::TextSize};
+
+use crate::{
+    lexer::{LexicalError, LexicalErrorType},
+    token::Tok,
+};
+
+/// Returns an error for invalid assignment targets.
+///
+/// # Errors
+///
+/// This returns an error when any of the given expressions are themselves
+/// or contain an expression that is invalid on the left hand side of an
+/// assignment. For example, all literal expressions are invalid assignment
+/// targets.
+pub(crate) fn assignment_targets(
+    targets: &[Expr],
+) -> Result<(), LalrpopError<TextSize, Tok, LexicalError>> {
+    for t in targets {
+        assignment_target(t)?;
+    }
+    Ok(())
+}
+
+/// Returns an error if the given target is invalid for the left hand side of
+/// an assignment.
+///
+/// # Errors
+///
+/// This returns an error when the given expression is itself or contains an
+/// expression that is invalid on the left hand side of an assignment. For
+/// example, all literal expressions are invalid assignment targets.
+pub(crate) fn assignment_target(
+    target: &Expr,
+) -> Result<(), LalrpopError<TextSize, Tok, LexicalError>> {
+    // Allowing a glob import here because of its limited scope.
+    #[allow(clippy::enum_glob_use)]
+    use self::Expr::*;
+
+    let err = |location: TextSize| -> LalrpopError<TextSize, Tok, LexicalError> {
+        let msg = "invalid assignment target";
+        let error = LexicalError {
+            error: LexicalErrorType::OtherError(msg.to_string()),
+            location,
+        };
+        LalrpopError::User { error }
+    };
+    match *target {
```

Comment:
Can I ask why you do `match *target` and then `BoolOp(ref e)`, instead of `match target` and `BoolOp(e)`? I'm curious :)

Reply:
Yes! I had written Rust for years before "[match ergonomics](https://rust-lang.github.io/rfcs/2005-match-ergonomics.html)" landed. That RFC essentially absolved you of needing to worry about using `ref` or `ref mut` or whatever. But I had always found the `ref` and `ref mut` bindings to be helpful for knowing when something was borrowed versus when ownership had been transferred.

It's a small matter overall. Do you want me to drop the `ref`? (I wonder if there is a Clippy lint for it.)

Reply:
No I don't mind it, I was just curious why, but that makes sense. Unless I really can't help it, I try to avoid asking for changes like this that we can't enforce automatically via tooling.

(Relatedly we often talk about how we want the box match thing from nightly.)

Reply:
> (Relatedly we often talk about how we want the box match thing from nightly.)

Yeah, that one has been unstable since the beginning sadly. I don't think think it's landing any time soon unfortunately. :-(

Reply:
We mostly want it in case we choose to `Box` our large AST variants, since we pattern match a lot!